content/article/deploying-dancer-apps-addendum.md

```diff
@@ -0,0 +1,164 @@
+    "categories"  : "tooling"
+  }
+
+*This article was originally published at
+[Perl Hacks](https://perlhacks.com/2024/08/deploying-dancer-apps-addendum/).*
+
+---
+
+Back in May, I wrote a blog post about how [I had moved a number of Dancer2
+applications](https://www.perl.com/article/deploying-dancer-apps/) to a new
+server and had, in the process, created a standardised procedure for
+deploying Dancer2 apps. It’s been about six weeks since I did that and I
+thought it would be useful to give a little update on how it all went and
+talk about a few little changes I’ve made.
+
+I mentioned that I was moving the apps to a new server. What I didn’t say was
+that I was convinced my old server was overpowered (and overpriced!) for what
+I needed, so the new server has less RAM and, I think, a slower CPU than the
+old one. And that turned out to be a bit of a problem. It turned out there
+was a time early each morning when there were too many requests coming into
+the server and it ran out of memory. I was waking up most days to a dead
+server. My previous work meant that fixing it wasn’t hard, but it really
+wasn’t something that I wanted to do each most mornings.
+
+So I wanted to look into reducing the amount of memory used by the apps. And
+that turned out to be a two-stage approach.
+
+You might recall that the apps were all controlled using a standardised
+driver program called “app_service”. It looked like this:
+
+```perl
+#!/usr/bin/env perl
+ 
+use warnings;
+use strict;
+use Daemon::Control;
+ 
+use ENV::Util -load_dotenv;
+ 
+use Cwd qw(abs_path);
+use File::Basename;
+ 
+Daemon::Control->new({
+  name      => ucfirst lc $ENV{KLORTHO_APP_NAME},
+  lsb_start => '$syslog $remote_fs',
+  lsb_stop  => '$syslog',
+  lsb_sdesc => 'Advice from Klortho',
+  lsb_desc  => 'Klortho knows programming. Listen to Klortho',
+  path      => abs_path($0),
+ 
+  program      => '/usr/bin/starman',
+  program_args => [ '--workers', 10, '-l', ":$ENV{KLORTHO_APP_PORT}",
+                    dirname(abs_path($0)) . '/app.psgi' ],
+ 
+  user  => $ENV{KLORTHO_OWNER},
+  group => $ENV{KLORTHO_GROUP},
+ 
+  pid_file    => "/var/run/$ENV{KLORTHO_APP_NAME}.pid",
+  stderr_file => "$ENV{KLORTHO_LOG_DIR}/error.log",
+  stdout_file => "$ENV{KLORTHO_LOG_DIR}/output.log",
+ 
+  fork => 2,
+})->run;
+```
+
+We’re deferring most of the clever stuff to
+[Daemon::Control](https://metacpan.org/pod/Daemon::Control). But we’re
+building the parameters to pass to the constructor. And two of the parameters
+(“program” and “program_args”) control how the service is run. You’ll see
+I’m using Starman. The first fix was obvious when you look at my code.
+Starman is a pre-forking server and we always start with 10 copies of the
+app. Now, I’m very proud of some of my apps, but I think it’s optimistic
+to think my Klortho server will ever need to respond to 10 simultaneous
+requests. Honestly, I’m pleasantly surprised if it gets 10 requests in a
```

Comment:
I'm guessing the static assets are served up by something other than the application? Might be helpful to mention that.

Reply:
I won't change that for the same reason.